algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala

```diff
@@ -0,0 +1,342 @@
+    (dropped.map(_.size).sum, remaining.reverse)
+  }
+
+  /**
+   * Converts the supplied buckets into a NEW vector of buckets
+   * satisfying this law:
+   *
+   * {{{
+   * rebucket(buckets, desired).map(_.size).sum == desired
+   * }}}
+   *
+   * (rebucket only works if desired.sum == buckets.map(_.size).sum)
+   *
+   * @param buckets [buckets] sorted in DESCENDING order (recent first)
+   * @param desired bucket sizes to rebucket `buckets` into.
+   */
+  private[algebird] def rebucket(buckets: Vector[Bucket], desired: Vector[Long]): Vector[Bucket] =
+    if (desired.isEmpty) Vector.empty
+    else {
+      val input = buckets.dropWhile(_.size == 0)
+      val bucketSize +: tail = desired
+      val remaining = drop(bucketSize, input)
+      input.head.copy(size = bucketSize) +: rebucket(remaining, tail)
+    }
+
+  /**
+   * @param toDrop total count to remove from the left of `input`.
+   * @param input buckets
+   * @return Vector with buckets, or pieces of buckets, with sizes
+   *         totalling `toDrop` items removed from the head. If an
+   *         element wasn't fully consumed, the remainder will be
+   *         stuck back onto the head.
+   */
+  @tailrec private[this] def drop(toDrop: Long, input: Vector[Bucket]): Vector[Bucket] = {
+    val (b @ Bucket(count, _)) +: tail = input
+    (toDrop - count) match {
+      case 0 => tail
+      case x if x < 0 => b.copy(size = -x) +: tail
+      case x if x > 0 => drop(x, tail)
+    }
+  }
+}
+
+/**
+ * @param conf the config values for this isntance.
+ * @param buckets Vector of timestamps of each (powers of 2)
+ *                ticks. This is the core of the exponential
+ *                histogram representation. See [[Canonical]] for
+ *                more info.
+ * @param total total ticks tracked, == buckets.map(_.size).sum
+ * @param time current timestamp of this instance.
+ *             Used with conf.windowSize to expire buckets.
+ */
+case class ExpHist(conf: ExpHist.Config, buckets: Vector[ExpHist.Bucket], total: Long, time: Long) {
+  import ExpHist.Bucket
+
+  /**
+   * Steps this instance forward to the new supplied time. Any
+   * buckets with a timestamp <= (newTime - conf.windowSize) will be
+   * evicted.
+   *
+   * @param newTime the new current time.
+   * @return ExpHist instance stepped forward to newTime.
+   */
+  def step(newTime: Long): ExpHist =
+    if (newTime <= time) this
+    else {
+      val (dropped, filtered) = conf.dropExpired(buckets, newTime)
+      copy(time = newTime, buckets = filtered, total = total - dropped)
+    }
+
+  /**
+   * Increment ExpHist by 1 at the supplied timestamp.
+   */
+  def inc(timestamp: Long): ExpHist = add(1L, timestamp)
+
+  /**
+   * Increment ExpHist by delta at the supplied timestamp.
+   */
+  def add(delta: Long, timestamp: Long): ExpHist = {
+    val self = step(timestamp)
+    if (delta == 0) self
+    else self.addAllWithoutStep(Vector(Bucket(delta, timestamp)), delta)
```

Comment:
here we can step BEFORE, because we know that there's only a single timestamp coming in.